### PR TITLE
[ENH] Add scale bar to plotting displays

### PR DESCRIPTION
**Pull request by oesteban**
Enables the annotation of plots with a scale bar:

nilearn-pr1
nilearn-pr2

- [x] All parameters can be passed on as `kwargs`
- [x] Scale can be above or below the label
- [x] Supports `'mm'` and `'cm'`

This addition is justified by the [requirements of journals](nilearn-pr1):

> Scale bars should be used rather than magnification factors, with the length of the bar defined in the legend rather than on the bar itself.

(their example):
nilearn-pr3

### CODE REVIEW

File: nilearn/plotting/displays.py [hunk 1]
```diff
@@ -137,6 +140,38 @@ def draw_left_right(self, size, bg_color, **kwargs):
...
+    def draw_scale_bar(self, bg_color, scale_width=5, units='cm', **kwargs):
+        from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar
+        ax = self.ax
+        fontproperties = kwargs.get('fontproperties', FontProperties())
+        if kwargs.get('size'):
+            fontproperties.set_size(kwargs.pop('size'))
+
+        # xlims = ax.get_xlim()
+        # print(xlims[1] - xlims[0])
+        # pixsize = (xlims[1] - xlims[0]) / self.shape[0]
```
Comment (bthirion) [inline nilearn/plotting/displays.py]: Please remove commented lines.

**Comment by oesteban** [comment]
:+1:

**Comment by oesteban** [comment]
I'm very unclear whether the test crash is related to this PR. Could someone with more experience take a look?

**Comment by KamalakerDadi** [comment]
>I'm very unclear whether the test crash is related to this PR. 

I pushed the rebuild to see if we can still get the same error.

**Comment by oesteban** [comment]
Any interest in adding some keyword to plot the annotation just in one of the panels of the mosaic? (i.e. typically one of those in corners).

File: nilearn/plotting/displays.py [hunk 2]
```diff
@@ -137,6 +140,35 @@ def draw_left_right(self, size, bg_color, **kwargs):
                 bbox=dict(boxstyle="square,pad=0", ec=bg_color, fc=bg_color),
                 **kwargs)
 
+    def draw_scale_bar(self, bg_color, scale_width=5, units='cm', **kwargs):
+        from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar
```
Comment (kchawla-pi) [inline nilearn/plotting/displays.py]: missing docstring

File: nilearn/plotting/displays.py [hunk 3]
```diff
@@ -137,6 +140,35 @@ def draw_left_right(self, size, bg_color, **kwargs):
...
+        ax = self.ax
+        fontproperties = kwargs.get('fontproperties', FontProperties())
```
Comment (kchawla-pi) [inline nilearn/plotting/displays.py]: fontproperties -> font_properties (easier to read and PEP8 compliant).
Same with
frameon -> frame_on

File: nilearn/plotting/displays.py [hunk 4]
```diff
@@ -889,6 +922,11 @@ def annotate(self, left_right=True, positions=True, size=12, **kwargs):
                 display_ax.draw_position(size=size, bg_color=bg_color,
                                          **kwargs)
 
+        if scalebar:
+            axes = self.axes.values()
+            for display_ax in axes:
```
Comment (kchawla-pi) [inline nilearn/plotting/displays.py]: display_ax -> display_axis
Saving 2 letters, not worth the obfuscation.

File: nilearn/plotting/displays.py [hunk 5]
```diff
@@ -16,6 +16,7 @@
 from matplotlib import lines
 from matplotlib import transforms, colors
```
Comment (kchawla-pi) [inline nilearn/plotting/displays.py]: PEP8 + alphabetical order of imports makes it easier to spot import one may be looking for.
```
from matplotlib import (colors,
                       lines,
                       transforms,
                       )

```

File: nilearn/plotting/displays.py [hunk 6]
```diff
@@ -16,6 +16,7 @@
...
 from matplotlib.colorbar import ColorbarBase
+from matplotlib.font_manager import FontProperties
 from scipy import sparse, stats
 
 from . import glass_brain, cm
```
Comment (kchawla-pi) [inline nilearn/plotting/displays.py]: Alphabetical order
` from . import cm, glass_brain`

File: nilearn/plotting/displays.py (hunk 3 above)
Comment (kchawla-pi) [reply nilearn/plotting/displays.py]: ax -> axis or axes depending on one axis or multplie

File: nilearn/plotting/displays.py [hunk 7]
```diff
@@ -137,6 +140,35 @@ def draw_left_right(self, size, bg_color, **kwargs):
...
+        fontproperties = kwargs.get('fontproperties', FontProperties())
+        if kwargs.get('size'):
+            fontproperties.set_size(kwargs.pop('size'))
+
+        length = scale_width
+        if units == 'cm':
+            length *= 10
+
+        frameon = kwargs.pop('frameon', False) is True
+        asb = AnchoredSizeBar(
```
Comment (kchawla-pi) [inline nilearn/plotting/displays.py]: `asb` -> `anchor_size_bar` is more self-documenting.

**Comment by GaelVaroquaux** [comment]
Sorry for the super slow turn around. I wanted to merge this PR: it looks like a super useful functionality.

However, with a cursory look at it, I wonder if the scale does not overlap with the cut coordinates by default. Is it the case?

**Comment by oesteban** [comment]
Hi, no worries @GaelVaroquaux for the turn around and thanks @kchawla-pi for the comments (will address them asap).

Please, let me re-check the implementation before I answer your question about overlap. I can remember I took that into account, but I want to give you an accurate reply.

**Comment by GaelVaroquaux** [comment]
Another comment: we have a rule in nilearn which is that every functionality should be demoed in an example. This is slightly related to the above, as it showed the default / simple usage of the functionality.

A simple way to do this would be to add a cell in http://nilearn.github.io/auto_examples/01_plotting/plot_demo_more_plotting.html#sphx-glr-auto-examples-01-plotting-plot-demo-more-plotting-py, maybe at the end.

Thanks!

And with regards to the turn around time, I feel bad, because we are trying to release, and I am giving you feedback only just when we are trying to release. It makes it hard for you to get in the next beta. My apologies!

**Comment by oesteban** [comment]
I see. I'll try to get this finished today so that it may be considered for the next beta :)

**Comment by GaelVaroquaux** [comment]
We'll try to give fast feedback

⁣Sent from my phone. Please forgive typos and briefness.​

**Comment by oesteban** [comment]
Checking whether the new auto_example I added works, right now. Will push when ready.

@GaelVaroquaux, I checked on the potential issue you mentioned. The scale bar is plotted on the bottom-right corner by default, so it should not overlap. It can be moved to any other position with the `loc` kwarg. The screen captures I gave as example are done that way (and also removing the position annotation when the scale bar is done). I could add a warning if both position and scale bar (with loc=3) are True.

File: nilearn/plotting/displays.py (hunk 2 above)
Comment (oesteban) [reply nilearn/plotting/displays.py]: (docstring has been added)

File: nilearn/plotting/displays.py (hunk 4 above)
Comment (oesteban) [reply nilearn/plotting/displays.py]: `display_ax` is being used many times in the previous code. I have no issue updating this instance, but that will make it inconsistent with the rest of the file.

Comment (GaelVaroquaux) [reply nilearn/plotting/displays.py]: Yes, I don't mind it.

**Comment by oesteban** [comment]
For some reason I don't understand, tests are failing.

I added the new examples here:

https://2-125426329-gh.circle-artifacts.com/0/home/circleci/project/doc/_build/html/auto_examples/01_plotting/plot_demo_more_plotting.html#annotating-plots

Please let me know if they get along with current documentation.

**Comment by GaelVaroquaux** [comment]
With regards to test failing, it's probably related to things external to nilearn (URLs changing, and some of our dependencies adding deprecation warnings). We are trying to fix this in parallel.

File: nilearn/plotting/displays.py [hunk 8]
```diff
@@ -137,6 +142,50 @@ def draw_left_right(self, size, bg_color, **kwargs):
...
+
+        frame_on = kwargs.pop('frameon', False) is True
+        anchor_size_bar = AnchoredSizeBar(
+            axis.transData,
+            width_mm,
+            '%g%s' % (scale_width, units),
+            loc=kwargs.pop('loc', 4),
+            pad=kwargs.pop('pad', 0.1),
+            borderpad=kwargs.pop('borderpad', 0.5),
+            sep=kwargs.pop('sep', 5),
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: These arguments should not be in the kwargs, but as explicit arguments to the function, with the documentation, as they are fairly central to its use.

Also, the docstring should document where the kwargs are sent, so that people can look up what kwargs are acceptable.

**Comment by GaelVaroquaux** [comment]
> It can be moved to any other position with the loc kwarg.

But the loc kwarg is currently documented nowhere, so it's hard for users
to figure this out.

File: nilearn/plotting/displays.py [hunk 9]
```diff
@@ -873,8 +923,7 @@ def annotate(self, left_right=True, positions=True, size=12, **kwargs):
         size: integer, optional
             The size of the text used.
         kwargs:
-            Extra keyword arguments are passed to matplotlib's text
-            function.
+            Extra keyword arguments are passed to matplotlib.
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: I really don't like it that it is now unclear where the kwargs are sent. It makes them very hard to use.

File: nilearn/plotting/displays.py [hunk 10]
```diff
@@ -884,15 +933,28 @@ def annotate(self, left_right=True, positions=True, size=12, **kwargs):
                 kwargs['color'] = 'k'
 
         bg_color = ('k' if self._black_bg else 'w')
+
+        # Cleanup scale bar's kwargs
+        scale_bar_kwargs = {
+            key: kwargs.pop(key) for key in ['scale_width', 'units']
+            if key in kwargs}
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: These arguments should be added explicitly to the function signature, rather than via the kwargs.

File: nilearn/plotting/displays.py [hunk 11]
```diff
@@ -859,7 +908,8 @@ def add_markers(self, marker_coords, marker_color='r', marker_size=30,
             display_ax.ax.scatter(xdata, ydata, s=marker_size,
                                   c=marker_color, **kwargs)
 
-    def annotate(self, left_right=True, positions=True, size=12, **kwargs):
+    def annotate(self, left_right=True, positions=True, scalebar=False,
+                 size=12, **kwargs):
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: The "scalebar" argument is not documented.

The documentation of the "scalebar" parameter should point to the "draw_scale_bar" for finer control.

File: nilearn/plotting/displays.py [hunk 12]
```diff
@@ -884,15 +933,28 @@ def annotate(self, left_right=True, positions=True, size=12, **kwargs):
...
         if positions:
-            for display_ax in self.axes.values():
-                display_ax.draw_position(size=size, bg_color=bg_color,
-                                         **kwargs)
+            for display_axis in self.axes.values():
+                display_axis.draw_position(size=size, bg_color=bg_color,
+                                           **kwargs)
+
+        if scalebar:
+            kwargs.update(scale_bar_kwargs)  # Insert kwargs back
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: kwargs should not be passed to draw_scale_bar. kwargs should be passed to only one type of function. If it's really necessary to route arguments down to this function, they should be added explicitly to the list of arguments of the calling function.

I am not convinced that extra arguments for the scalebar need to be passed around. Users can always use the specific function.
I am not convinced that it is

**Comment by oesteban** [comment]
I hope I addressed all your comments, @GaelVaroquaux. Thanks very much for the feedback.

File: nilearn/plotting/displays.py [hunk 13]
```diff
@@ -137,6 +142,75 @@ def draw_left_right(self, size, bg_color, **kwargs):
...
+            width_mm,
+            '%g%s' % (scale_width, units),
+            color=color,
+            loc=scalebar_kwargs.pop('loc', 4),
+            pad=scalebar_kwargs.pop('pad', 0.1),
+            borderpad=scalebar_kwargs.pop('borderpad', 0.5),
+            sep=scalebar_kwargs.pop('sep', 5),
+            frameon=frame_on,
+            fontproperties=font_properties,
+            **scalebar_kwargs)
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: It is bad practice to pop some kwargs and pass others. There is no logic, and it forces the user to read the code of the function to know what the defaults are. The kwargs that you pass should simply be arguments of the function: the signature of the function should be:
<pre>
draw_scale_bar(self, bg_color, ..., loc=4, pad=.1, borderpad=.5, sep=5, color=...)
</pre>

File: nilearn/plotting/displays.py [hunk 14]
```diff
@@ -137,6 +142,75 @@ def draw_left_right(self, size, bg_color, **kwargs):
...
+            bar annotation);
+            `sep` (float - separation between text and the bar);
+            for additional keywords please check matplotlib's
+            `AnchoredSizeBar <https://matplotlib.org/mpl_toolkits/axes_grid\
+            /api/anchored_artists_api.html>`__ documentation.
+        kwargs:
+            Extra keyword arguments are passed to matplotlib's
+            FontProperties object used by AnchoredSizeBar.
+        """
+        from mpl_toolkits.axes_grid1.anchored_artists import AnchoredSizeBar
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: Why is this import delayed? (just asking, there might be a good reason).

File: nilearn/plotting/displays.py [hunk 15]
```diff
@@ -137,6 +142,75 @@ def draw_left_right(self, size, bg_color, **kwargs):
...
+
+        # Sanity check to replace None by empty dict
+        scalebar_kwargs = scalebar_kwargs or {}
+
+        axis = self.ax
+
+        color = kwargs.pop('color')
+        font_properties = scalebar_kwargs.get('fontproperties',
+                                              FontProperties(**kwargs))
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: This dance of kwargs makes me nervous. It's really hard to follow where arguments are routed even when reading the code. Hence for an end user, it is going to be guesswork.

File: nilearn/plotting/displays.py [hunk 16]
```diff
@@ -870,8 +945,13 @@ def annotate(self, left_right=True, positions=True, size=12, **kwargs):
         positions: boolean, optional
             If positions is True, annotations indicating the
             positions of the cuts are drawn.
+        scalebar: boolean, optional
+            If scalebar is True, cuts are annotated with a reference
+            scale bar (bottom-right position by default).
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: Here you should reference the method  "draw_scale_bar" for finer control of the scale bar.

File: nilearn/plotting/displays.py [hunk 17]
```diff
@@ -870,8 +945,13 @@ def annotate(self, left_right=True, positions=True, size=12, **kwargs):
...
         size: integer, optional
             The size of the text used.
+        scalebar_kwargs: dict, optional
+            Arguments to control how the scale bar is drawn.
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: I would remove these and point to the method "draw_scale_bar" if people want finer control.

Comment (GaelVaroquaux) [reply nilearn/plotting/displays.py]: One option is to extract the most import options and add them here explicitly, eg "scalebar_units", "scalebar_pos". The reason that I am insisting is that "kwargs" are really not user friendly for non experts.

File: nilearn/plotting/displays.py (hunk 11 above)
Comment (oesteban) [reply nilearn/plotting/displays.py]: Done

**Comment by oesteban** [comment]
I hope I addressed all your comments, @GaelVaroquaux. Thanks very much for the feedback.

File: nilearn/plotting/displays.py [hunk 18]
```diff
@@ -870,11 +950,17 @@ def annotate(self, left_right=True, positions=True, size=12, **kwargs):
...
+            For finer control of the scale bar, please check out
+            :meth:`~nilearn.plotting.displays.BaseAxes.draw_scale_bar`.
         size: integer, optional
             The size of the text used.
         kwargs:
             Extra keyword arguments are passed to matplotlib's text
-            function.
+            function. If the keyword starts with `scale_`, then it is
+            redirected to :meth:`~nilearn.plotting.displays.BaseAxes.\
+            draw_scale_bar` if ``scalebar`` is ``True``.
```
Comment (GaelVaroquaux) [inline nilearn/plotting/displays.py]: I am not sure how to say this: important argument should not be in kwargs. The function that you are contributing is not usable without understanding the kwargs. Understanding them requires careful reading of several docstrings. This is bad user experience.

An important goal of nilearn is to minimize cognitive overhead for users. This is why people like it: non experts can use it. Code that is contributed to nilearn should abide by these standard. I would say that the main goal of contributing code to nilearn should be to put it in the hand of non experts. Elsewhere code could live in a separate repository.

As I have said, the useful arguments should be passed as explicit arguments and not in kwargs.

I will not merge the code as such, as the goal of development in nilearn is to improvement usability at each new release.

Comment (GaelVaroquaux) [reply nilearn/plotting/displays.py]: By the way, if this is helpful, I can implement this change, if I am not conveying my ideas clearly or if you don't have time.

Comment (oesteban) [reply nilearn/plotting/displays.py]: This got hidden from my email:

I just read the last comment about making the changes yourself. Of course,
go for it! I'd really appreciate that and will probably save time to both
of us.

Thanks very much

**Comment by GaelVaroquaux** [comment]
@oesteban : sorry for being unclear with my comments, and thanks for being flexible. I'll do a PR to your branch ASAP.

**Comment by GaelVaroquaux** [comment]
I've pushed directly to your branch. Sorry, I wanted to do a PR, but I don't know why github wouldn't let me do that.

**Comment by oesteban** [comment]
No worries

**Comment by GaelVaroquaux** [comment]
@oesteban : are you fine with the changes? If so, I'll merge when CI is green.

**Comment by bthirion** [comment]
LGTM. Hopefully tests will pass...

**Comment by oesteban** [comment]
Sure! Let's do this

**Comment by GaelVaroquaux** [comment]
CircleCI found a bad reference. I fixed it, and we are in for a new round of computation.

**Comment by GaelVaroquaux** [comment]
CI ran. Merging
